src/VectorArrows3d.ts

```diff
@@ -0,0 +1,349 @@
+import BaseDrawableMeshObject from "./BaseDrawableMeshObject";
+import { MESH_NO_PICK_OCCLUSION_LAYER } from "./ThreeJsPanel";
+
+const DEFAULT_DIAMETER = 0.002;
+
+// Unscaled arrowhead dimensions. These will be scaled by the diameter.
+const HEAD_BASE_RADIUS = 1.5;
+const HEAD_BASE_HEIGHT = 4;
+
+const DEFAULT_INSTANCE_COUNT = 256;
+
+/**
+ * A drawable vector arrow field, which uses instanced meshes for performance.
+ */
+export default class VectorArrows3d extends BaseDrawableMeshObject implements IDrawableObject {
+  /**
+   * Scale of this object in world coordinates, when unscaled. Used to
+   * compensate for parent transforms in order to keep arrow meshes from being
+   * distorted.
+   */
+  protected worldScale: Vector3;
+
+  private headInstancedMesh: InstancedMesh;
+  private shaftInstancedMesh: InstancedMesh;
+
+  private maxInstanceCount: number;
+  private positions: Float32Array | null;
+  private deltas: Float32Array | null;
+  private colors: Float32Array | null;
+  private diameter: Float32Array;
+
+  // Temporary calculation objects. Optimization taken from three.js examples.
+  private tempDst: Vector3;
+  private tempScale: Vector3;
+  private tempMatrix: Object3D;
+
+  constructor() {
+    super();
+    this.worldScale = new Vector3(1, 1, 1);
+
+    this.meshPivot.layers.set(MESH_NO_PICK_OCCLUSION_LAYER);
+
+    this.maxInstanceCount = DEFAULT_INSTANCE_COUNT;
+    const { headMesh: headMesh, shaftMesh: shaftMesh } = this.initInstancedMeshes(DEFAULT_INSTANCE_COUNT);
+    this.headInstancedMesh = headMesh;
+    this.shaftInstancedMesh = shaftMesh;
+    this.headInstancedMesh.count = 0;
+    this.shaftInstancedMesh.count = 0;
+
+    this.positions = null;
+    this.deltas = null;
+    this.colors = null;
+    this.diameter = new Float32Array([DEFAULT_DIAMETER]);
+
+    this.tempDst = new Vector3();
+    this.tempScale = new Vector3();
+    this.tempMatrix = new Object3D();
+
+    this.onParentTransformUpdated();
+  }
+
+  /**
+   * Returns (unscaled) buffer geometry for the head and shaft parts of the
+   * arrow.
+   * @returns
+   * - `head`: BufferGeometry for the arrowhead, a cone pointing along the +Z
+   *   axis, with the pivot at the tip of the cone. Height and radius are based
+   *   on constant values (`HEAD_BASE_HEIGHT` and `HEAD_BASE_RADIUS`).
+   * - `shaft`: BufferGeometry for the cylindrical arrow shaft. The cylinder
+   *   points along the +Z axis, with the pivot at the base of the cylinder.
+   *   Height is 1 and diameter is 1.
+   *
+   * ```txt
+   *  ^ +Z axis ^
+   *  _____      x
+   * |     |    / \
+   * |     |   /   \
+   *  --x--   /-----\
+   *
+   *   x = pivot (0,0,0)
+   * ```
+   */
+  private static generateGeometry(): { head: BufferGeometry; shaft: BufferGeometry } {
+    // TODO: Currently the shape of the arrow head is fixed. Allow configuring
+    // this in the future?
+    const cylinderGeometry = new CylinderGeometry(0.5, 0.5, 1, 8, 1, false);
```

Comment:
nit: can't help but think DEFAULT_DIAMETER should be showing up in here.  but mostly, the ConeGeometry below uses symbolic constants and this one doesn't.